server.go

```diff
@@ -320,16 +320,20 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 			}
 			return err
 		}
-		if traceInfo.tr != nil {
-			// TODO: set payload.msg to something that
-			// prints usefully with %s; req is a []byte.
-			traceInfo.tr.LazyLog(&payload{sent: false}, true)
-		}
 		switch pf {
 		case compressionNone:
 			statusCode := codes.OK
 			statusDesc := ""
-			reply, appErr := md.Handler(srv.server, ctx, s.opts.codec, req)
+			df := func(b []byte, v interface{}) error {
```

Comment:
I think one part of the issue is that we want to unmarshal the request, and log that before going anywhere near the method handler. But this code doesn't know how to unmarshal the request because it doesn't know how to construct the destination type for passing to codec.Unmarshal.

The streaming equivalent (stream.go:RecvMsg) works by taking the message to be populated. The generated stream handler calls RecvMsg, though that works because a stream starts and _then_ you deal with the incoming message, whereas a unary RPC couples those two steps. It works for streams because there's a value specific to the stream (the `Service_MethodServer` interface embeds a `grpc.ServerStream` that the handler can use).


Maybe there's no way around it. It still feels wrong to have a `decodeFunc`, which explicitly here does more than a "decode". It's a hack to get the message traced between unmarshaling and running the server's unary RPC handler. I don't have a better solution right now, but it might be good to think about this a bit before we make a change.